File: api/claude.py
from http.server import BaseHTTPRequestHandler
import json
import sys
import traceback
import anyio
from claude_code_sdk import query, ClaudeCodeOptions, AssistantMessage, TextBlock

class handler(BaseHTTPRequestHandler):
    def do_GET(self):
        """GET リクエスト処理 - API 状態確認"""
        self.send_response(200)
        self.send_header('Content-Type', 'application/json')
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type')
        self.send_header('Access-Control-Allow-Methods', 'GET, POST, OPTIONS')
        self.end_headers()
        
        response = {
            "status": "success",
            "message": "Claude Code SDK API is running",
            "endpoint": "/api/claude",
            "version": "1.0.0",
            "python_version": sys.version,
            "methods": ["GET", "POST"]
        }
        
        self.wfile.write(json.dumps(response, ensure_ascii=False).encode('utf-8'))
    
    def do_POST(self):
        """POST リクエスト処理 - Claude Code SDK クエリ実行"""
        try:
            # リクエストボディ読み取り
            content_length = int(self.headers.get('Content-Length', 0))
            post_data = self.rfile.read(content_length).decode('utf-8')
            request_data = json.loads(post_data)
            
            # 必須パラメータチェック
            if 'prompt' not in request_data:
                self._send_error_response(400, "Missing required parameter: prompt")
                return
            
            prompt = request_data['prompt']
            
            # Claude Code SDK クエリ実行
            response_text = anyio.run(self._query_claude, prompt)
            
            # 成功レスポンス
            self._send_success_response({
                "status": "success",
                "prompt": prompt,
                "response": response_text,
                "timestamp": self._get_timestamp()
            })
            
        except json.JSONDecodeError:
            self._send_error_response(400, "Invalid JSON in request body")
        except Exception as e:
            self._send_error_response(500, f"Internal server error: {str(e)}")
    
    async def _query_claude(self, prompt):
        """Claude Code SDK クエリ実行（非同期）"""
        try:
            response_parts = []
            
            # Claude Code SDK クエリ実行
            async for message in query(prompt=prompt):
                if isinstance(message, AssistantMessage):
                    for block in message.content:
                        if isinstance(block, TextBlock):
                            response_parts.append(block.text)
            
            return "".join(response_parts)
            
        except Exception as e:
            print(f"Claude SDK Error: {e}")
            traceback.print_exc()
            raise e
    
    def do_OPTIONS(self):
        """CORS preflight リクエスト処理"""
        self.send_response(200)
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type')
        self.send_header('Access-Control-Allow-Methods', 'GET, POST, OPTIONS')
        self.end_headers()
    
    def _send_success_response(self, data):
        """成功レスポンス送信"""
        self.send_response(200)
        self.send_header('Content-Type', 'application/json')
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type')
        self.send_header('Access-Control-Allow-Methods', 'GET, POST, OPTIONS')
        self.end_headers()
        
        self.wfile.write(json.dumps(data, ensure_ascii=False).encode('utf-8'))
    
    def _send_error_response(self, status_code, message):
        """エラーレスポンス送信"""
        self.send_response(status_code)
        self.send_header('Content-Type', 'application/json')
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type')
        self.send_header('Access-Control-Allow-Methods', 'GET, POST, OPTIONS')
        self.end_headers()
        
        error_response = {
            "status": "error",
            "message": message,
            "timestamp": self._get_timestamp()
        }
        
        self.wfile.write(json.dumps(error_response, ensure_ascii=False).encode('utf-8'))
    
    def _get_timestamp(self):
        """現在のタイムスタンプ取得"""
        from datetime import datetime
        return datetime.now().isoformat()

File: api/claude.ts
import { VercelRequest, VercelResponse } from '@vercel/node';
import { query, type SDKMessage } from '@anthropic-ai/claude-code';

export default async function handler(req: VercelRequest, res: VercelResponse) {
  // CORS設定
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Headers', 'Content-Type');
  res.setHeader('Access-Control-Allow-Methods', 'GET, POST, OPTIONS');

  // Preflightリクエストの処理
  if (req.method === 'OPTIONS') {
    res.status(200).end();
    return;
  }

  // GETリクエストの処理 - API状態確認
  if (req.method === 'GET') {
    res.status(200).json({
      status: 'success',
      message: 'Claude Code SDK API is running (TypeScript)',
      endpoint: '/api/claude',
      version: '2.0.0',
      runtime: 'Node.js',
      nodeVersion: process.version,
      methods: ['GET', 'POST']
    });
    return;
  }

  // POSTリクエストの処理 - Claude Code SDKクエリ実行
  if (req.method === 'POST') {
    try {
      // リクエストボディの検証
      const { prompt } = req.body;
      
      if (!prompt || typeof prompt !== 'string') {
        res.status(400).json({
          status: 'error',
          message: 'Missing required parameter: prompt',
          timestamp: new Date().toISOString()
        });
        return;
      }

      // Claude Code SDKクエリ実行
      console.log(`Processing query: ${prompt}`);
      const messages: SDKMessage[] = [];

      try {
        // 環境変数チェック
        if (!process.env.ANTHROPIC_API_KEY) {
          throw new Error('ANTHROPIC_API_KEY environment variable is not set');
        }

        // クエリ実行
        for await (const message of query({
          prompt: prompt,
          abortController: new AbortController(),
          options: {
            maxTurns: 1, // 1ターンのみ（安全性のため）
            apiKey: process.env.ANTHROPIC_API_KEY,
          }
        })) {
          messages.push(message);
        }

        // メッセージからテキストを抽出
        const responseText = messages
          .filter((msg: any) => msg.type === 'message' && msg.role === 'assistant')
          .map((msg: any) => {
            // メッセージ内容からテキストを抽出
            if (msg.content && Array.isArray(msg.content)) {
              return msg.content
                .filter((block: any) => block.type === 'text')
                .map((block: any) => block.text || '')
                .join('');
            }
            return '';
          })
          .join('');

        // 成功レスポンス
        res.status(200).json({
          status: 'success',
          prompt: prompt,
          response: responseText,
          timestamp: new Date().toISOString()
        });

      } catch (sdkError: any) {
        console.error('Claude SDK Error:', sdkError);
        res.status(500).json({
          status: 'error',
          message: `Claude SDK error: ${sdkError.message || 'Unknown error'}`,
          timestamp: new Date().toISOString()
        });
      }

    } catch (error: any) {
      console.error('Request processing error:', error);
      res.status(500).json({
        status: 'error',
        message: `Internal server error: ${error.message || 'Unknown error'}`,
        timestamp: new Date().toISOString()
      });
    }
    return;
  }

  // その他のメソッドは許可しない
  res.status(405).json({
    status: 'error',
    message: 'Method not allowed'
  });
          }
